### DISCUSSION TITLE
Math.gl Objects on Shaders

### DISCUSSION THREAD

**Discussion by ilan-gold** [visgl/deck.gl#5670] [Q&A] [answered]
Hello,

Can we pass in arbitrary Math.gl objects to the shaders like a [Plane](https://math.gl/modules/culling/docs/api-reference/plane) without transformation?  I tried it and it did not seem to work, even when I defined a `struct`.  Maybe I am missing something though.  I think the reason I am asking is that I got a "bad uniform" error when trying to pass in an array of `Vector3` objects as well, which made me think that it is not in general possible (I had to call `.flat()` on the array).  Also sorry for asking here but luma/math.gl don't have Discussion boards.

Thanks!

**Comment by ibgreen** [comment, marked as answer]
Basic math objects from `@math.gl/core` are subclasses of JavaScript Arrays and can be used directly as uniforms. 

But the objects in e.g. `@math.gl/culling` are more complex and are standard JavaScript classes. However, the have fields that are basic Vectors and numbers and (plane.normal and plane.distance in the case of Plane) and those can indeed be used directly with luma.gl

**Reply by ilan-gold** [reply]
Ok, that makes sense.  Should I be able to pass in an array of `Vector3` "objects" (`Vector3[]`) to the shader without calling `.flat()` on the array?  Maybe I was doing something wrong.

**Reply by ibgreen** [reply]
No that will not work without the `.flat`.

**Reply by ilan-gold** [reply]
Thanks!
